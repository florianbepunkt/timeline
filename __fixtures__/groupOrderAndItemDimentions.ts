export const orderedGroups = {
  1: {
    group: {
      id: "1",
    },
    index: 0,
  },
  2: {
    group: {
      id: "2",
    },
    index: 1,
  },
  3: {
    group: {
      id: "3",
    },
    index: 2,
  },
};

export const dimensionItems = [
  {
    dimensions: {
      collisionLeft: 1540540000000,
      collisionWidth: 6803877,
      height: 22.5,
      left: 907.4074074074074,
      order: {
        group: {
          id: "1",
        },
        index: 0,
      },
      stack: true,
      top: 7.5,
      width: 78.74857638888886,
    },
    id: "0",
  },
  {
    dimensions: {
      collisionLeft: 1540532800000,
      collisionWidth: 21203877,
      height: 22.5,
      left: 824.074074074074,
      order: {
        group: {
          id: "1",
        },
        index: 0,
      },
      stack: true,
      top: 37.5,
      width: 245.4152430555556,
    },
    id: "5",
  },
  {
    dimensions: {
      collisionLeft: 1540550800000,
      collisionWidth: 24803877,
      height: 22.5,
      left: 1032.4074074074074,
      order: {
        group: {
          id: "1",
        },
        index: 0,
      },
      stack: true,
      top: 7.5,
      width: 287.08190972222224,
    },
    id: "6",
  },
  {
    dimensions: {
      collisionLeft: 1540570000000,
      collisionWidth: 14875919,
      height: 22.5,
      left: 1254.6296296296296,
      order: {
        group: {
          id: "1",
        },
        index: 0,
      },
      stack: true,
      top: 37.5,
      width: 172.1749884259259,
    },
    id: "1",
  },
  {
    dimensions: {
      collisionLeft: 1540620000000,
      collisionWidth: 20397548,
      height: 22.5,
      left: 1833.3333333333333,
      order: {
        group: {
          id: "1",
        },
        index: 0,
      },
      stack: true,
      top: 7.5,
      width: 236.08273148148123,
    },
    id: "2",
  },
  {
    dimensions: {
      collisionLeft: 1540656000000,
      collisionWidth: 20397548,
      height: 22.5,
      left: 2250,
      order: {
        group: {
          id: "3",
        },
        index: 2,
      },
      stack: true,
      top: 105,
      width: 236.08273148148146,
    },
    id: "3",
  },
];
