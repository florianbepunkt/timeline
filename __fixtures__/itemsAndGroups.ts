import type { TimelineGroupBase, TimelineItemBase } from "../src";

export const items: TimelineItemBase[] = [
  {
    id: "0",
    group: "1",
    startTime: 1540540000000,
    endTime: 1540546803877,
    canMove: true,
    canResize: "both",
  },
  {
    id: "5",
    group: "1",
    startTime: 1540532800000,
    endTime: 1540554003877,
    canMove: true,
    canResize: "both",
    className: "",
  },
  {
    id: "6",
    group: "1",
    startTime: 1540550800000,
    endTime: 1540575603877,
    canMove: true,
    canResize: "both",
    className: "",
  },
  {
    id: "1",
    group: "1",
    startTime: 1540570000000,
    endTime: 1540584875919,
    canMove: true,
    canResize: "both",
  },
  {
    id: "2",
    group: "1",
    startTime: 1540620000000,
    endTime: 1540640397548,
    canMove: true,
    canResize: "both",
    className: "",
  },
  {
    id: "3",
    group: "3",
    startTime: 1540656000000,
    endTime: 1540676397548,
    canMove: true,
    canResize: "both",
    className: "",
  },
];

export const groups: TimelineGroupBase[] = [
  { id: "1", title: "1" },
  { id: "2", title: "2" },
  { id: "3", title: "3" },
];
