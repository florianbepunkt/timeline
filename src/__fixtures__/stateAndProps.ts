import { items } from "./itemsAndGroups";

export const visibleTimeStart = 1540501200000;
export const visibleTimeEnd = 1540587600000;

export const props = {
  lineHeight: 30,
  stackItems: true,
  itemHeightRatio: 0.75,
  visibleTimeEnd,
  visibleTimeStart,
};

export const propsNoStack = {
  ...props,
  stackItems: false,
};

export const state = {
  draggingItem: undefined,
  dragTime: null,
  resizingItem: null,
  resizingEdge: null,
  resizeTime: null,
  newGroupOrder: null,
  canvasTimeStart: 1540414800000,
  visibleTimeEnd: visibleTimeEnd,
  visibleTimeStart: visibleTimeStart,
  canvasTimeEnd: 1540674000000,
  width: 1000,
};

//offset 1 hour
const timeOffset = 1 * 60 * 60 * 1000;

export const stateMoveItem = {
  ...state,
  draggingItem: items[0].id,
  dragTime: items[0].startTime + timeOffset,
  newGroupOrder: 0,
};
export const stateResizeItemLeft = {
  ...state,
  resizingItem: items[0].id,
  resizingEdge: "left",
  resizeTime: items[0].startTime + timeOffset,
  newGroupOrder: 0,
};

export const stateResizeItemRight = {
  ...state,
  resizingItem: items[0].id,
  resizingEdge: "right",
  resizeTime: items[0].endTime + timeOffset,
  newGroupOrder: 0,
};
