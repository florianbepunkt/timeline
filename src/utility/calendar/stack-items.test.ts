import { describe, expect, test } from "vitest";
import { items, groups } from "../../../__fixtures__/itemsAndGroups";
import {
  props,
  propsNoStack,
  state,
  stateMoveItem,
  stateResizeItemLeft,
  stateResizeItemRight,
} from "../../../__fixtures__/stateAndProps";
import { stackTimelineItems } from "./calendar";
import type { TimelineItemEdge } from "../../types";

describe("stackItems", () => {
  test("work as expected", () => {
    expect(
      stackTimelineItems({
        calculateExtraSpace: false,
        canvasTimeEnd: state.canvasTimeEnd,
        canvasTimeStart: state.canvasTimeStart,
        canvasWidth: 9000,
        draggingItem: state.draggingItem ?? null,
        dragTime: state.dragTime,
        groups,
        itemHeight: props.itemHeightRatio,
        items,
        lineHeight: props.lineHeight,
        newGroupOrder: state.newGroupOrder,
        resizeTime: state.resizeTime,
        resizingEdge: state.resizingEdge,
        resizingItem: state.resizingItem,
        stackItems: props.stackItems,
      })
    ).toMatchSnapshot();
  });

  test("work as expected no stack", () => {
    expect(
      stackTimelineItems({
        calculateExtraSpace: false,
        canvasTimeEnd: state.canvasTimeEnd,
        canvasTimeStart: state.canvasTimeStart,
        canvasWidth: 9000,
        draggingItem: state.draggingItem ?? null,
        dragTime: state.dragTime,
        groups,
        itemHeight: propsNoStack.itemHeightRatio,
        items,
        lineHeight: propsNoStack.lineHeight,
        newGroupOrder: state.newGroupOrder,
        resizeTime: state.resizeTime,
        resizingEdge: state.resizingEdge,
        resizingItem: state.resizingItem,
        stackItems: propsNoStack.stackItems,
      })
    ).toMatchSnapshot();
  });

  test("should stack items while moving an item", () => {
    expect(
      stackTimelineItems({
        calculateExtraSpace: false,
        canvasTimeEnd: stateMoveItem.canvasTimeEnd,
        canvasTimeStart: stateMoveItem.canvasTimeStart,
        canvasWidth: 9000,
        draggingItem: stateMoveItem.draggingItem,
        dragTime: stateMoveItem.dragTime,
        groups,
        itemHeight: props.itemHeightRatio,
        items,
        lineHeight: props.lineHeight,
        newGroupOrder: stateMoveItem.newGroupOrder,
        resizeTime: stateMoveItem.resizeTime,
        resizingEdge: stateMoveItem.resizingEdge,
        resizingItem: stateMoveItem.resizingItem,
        stackItems: props.stackItems,
      })
    ).toMatchSnapshot();
  });

  test("should stack items while resize item left", () => {
    expect(
      stackTimelineItems({
        calculateExtraSpace: false,
        items,
        groups,
        canvasWidth: 9000,
        canvasTimeStart: stateResizeItemLeft.canvasTimeStart,
        canvasTimeEnd: stateResizeItemLeft.canvasTimeEnd,
        lineHeight: props.lineHeight,
        itemHeight: props.itemHeightRatio,
        stackItems: props.stackItems,
        draggingItem: stateResizeItemLeft.draggingItem ?? null,
        resizingItem: stateResizeItemLeft.resizingItem,
        dragTime: stateResizeItemLeft.dragTime,
        resizingEdge: stateResizeItemLeft.resizingEdge as TimelineItemEdge,
        resizeTime: stateResizeItemLeft.resizeTime,
        newGroupOrder: stateResizeItemLeft.newGroupOrder,
      })
    ).toMatchSnapshot();
  });

  test("should stack items while resize item right", () => {
    expect(
      stackTimelineItems({
        calculateExtraSpace: false,
        canvasTimeEnd: stateResizeItemRight.canvasTimeEnd,
        canvasTimeStart: stateResizeItemRight.canvasTimeStart,
        canvasWidth: 9000,
        draggingItem: stateResizeItemRight.draggingItem ?? null,
        dragTime: stateResizeItemRight.dragTime,
        groups,
        itemHeight: props.itemHeightRatio,
        items,
        lineHeight: props.lineHeight,
        newGroupOrder: stateResizeItemRight.newGroupOrder,
        resizeTime: stateResizeItemRight.resizeTime,
        resizingEdge: stateResizeItemRight.resizingEdge as TimelineItemEdge,
        resizingItem: stateResizeItemRight.resizingItem,
        stackItems: props.stackItems,
      })
    ).toMatchSnapshot();
  });

  test("should return empty dimensions if groups are empty", () => {
    expect(
      stackTimelineItems({
        calculateExtraSpace: false,
        items,
        groups: [],
        canvasWidth: 9000,
        canvasTimeStart: state.canvasTimeStart,
        canvasTimeEnd: state.canvasTimeEnd,
        lineHeight: props.lineHeight,
        itemHeight: props.itemHeightRatio,
        stackItems: props.stackItems,
        draggingItem: state.draggingItem ?? null,
        resizingItem: state.resizingItem,
        dragTime: state.dragTime,
        resizingEdge: state.resizingEdge,
        resizeTime: state.resizeTime,
        newGroupOrder: state.newGroupOrder,
      })
    ).toMatchObject({
      dimensionItems: [],
      height: 0,
      groupHeights: [],
      groupTops: [],
    });
  });
});
