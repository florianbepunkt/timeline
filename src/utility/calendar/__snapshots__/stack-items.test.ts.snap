// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stackItems > should stack items while moving an item 1`] = `
{
  "dimensionItems": [
    {
      "dimensions": {
        "collisionLeft": 1540543600000,
        "collisionWidth": 6803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4472.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 236.24572916666602,
      },
      "id": "0",
    },
    {
      "dimensions": {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4097.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    {
      "dimensions": {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4722.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    {
      "dimensions": {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 5388.888888888889,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    {
      "dimensions": {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 7125,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    {
      "dimensions": {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 8375,
        "order": {
          "group": {
            "id": "3",
            "title": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 104.625,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": [
    60,
    30,
    30,
  ],
  "groupTops": [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems > should stack items while resize item left 1`] = `
{
  "dimensionItems": [
    {
      "dimensions": {
        "collisionLeft": 1540543600000,
        "collisionWidth": 3203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4472.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 111.24572916666602,
      },
      "id": "0",
    },
    {
      "dimensions": {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4097.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    {
      "dimensions": {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4722.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    {
      "dimensions": {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 5388.888888888889,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    {
      "dimensions": {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 7125,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    {
      "dimensions": {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 8375,
        "order": {
          "group": {
            "id": "3",
            "title": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 104.625,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": [
    60,
    30,
    30,
  ],
  "groupTops": [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems > should stack items while resize item right 1`] = `
{
  "dimensionItems": [
    {
      "dimensions": {
        "collisionLeft": 1540540000000,
        "collisionWidth": 10403877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4347.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 361.245729166666,
      },
      "id": "0",
    },
    {
      "dimensions": {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4097.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    {
      "dimensions": {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4722.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    {
      "dimensions": {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 5388.888888888889,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    {
      "dimensions": {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 7125,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    {
      "dimensions": {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 8375,
        "order": {
          "group": {
            "id": "3",
            "title": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 104.625,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": [
    60,
    30,
    30,
  ],
  "groupTops": [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems > work as expected 1`] = `
{
  "dimensionItems": [
    {
      "dimensions": {
        "collisionLeft": 1540540000000,
        "collisionWidth": 6803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4347.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 236.24572916666602,
      },
      "id": "0",
    },
    {
      "dimensions": {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4097.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    {
      "dimensions": {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4722.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    {
      "dimensions": {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 5388.888888888889,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 44.625,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    {
      "dimensions": {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 7125,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    {
      "dimensions": {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 8375,
        "order": {
          "group": {
            "id": "3",
            "title": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 104.625,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": [
    60,
    30,
    30,
  ],
  "groupTops": [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems > work as expected no stack 1`] = `
{
  "dimensionItems": [
    {
      "dimensions": {
        "collisionLeft": 1540540000000,
        "collisionWidth": 6803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4347.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 236.24572916666602,
      },
      "id": "0",
    },
    {
      "dimensions": {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4097.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    {
      "dimensions": {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 4722.222222222223,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    {
      "dimensions": {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 5388.888888888889,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    {
      "dimensions": {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 7125,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 14.625,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    {
      "dimensions": {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": 0.75,
        "left": 8375,
        "order": {
          "group": {
            "id": "3",
            "title": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 74.625,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": [
    30,
    30,
    30,
  ],
  "groupTops": [
    0,
    30,
    60,
  ],
  "height": 90,
}
`;
